### PR TITLE
[CFE-33,34]: Feat - validator details overview section

### PR DESCRIPTION
**Pull request by Poafs1**
- [CFE-33](https://linear.app/alleslabs/issue/CFE-33/users-can-see-validator-top-section-information)
- [CFE-34](https://linear.app/alleslabs/issue/CFE-34/[overview]-users-see-validator-description)

### CODE REVIEW

File: src/lib/types/validator.ts [hunk 1]
```diff
@@ -22,7 +23,7 @@ export type Validator = z.infer<typeof zValidator>;
 export const zValidatorData = z
   .object({
     rank: z.number().nullable(),
-    rank_cummulative_voting_power: zBig,
+    rank_cummulative_voting_power: zBig.optional(),
```
Comment (evilpeach) [inline src/lib/types/validator.ts:26]: Let's patch to the newest version that add percent 33 and 66.

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: src/lib/components/ValidatorBadge.tsx [hunk 2]
```diff
@@ -43,24 +25,13 @@ export const ValidatorBadge = ({
...
-            <Image
-              boxSize={badgeSize}
-              src={valImgSrc}
-              alt={validator.moniker}
-              borderRadius="50%"
-              fallbackSrc="https://assets.alleslabs.dev/webapp-assets/placeholder/validator.svg"
-              fallbackStrategy="beforeLoadOrError"
-            />
-          )}
+          <ValidatorImage validator={validator} boxSize={badgeSize} />
```
Comment (songwongtp) [inline src/lib/components/ValidatorBadge.tsx:34]: Move validator image outside of the condition

File: src/lib/pages/validator-details/components/bonded-token-changes/VotingPowerChart.tsx [hunk 3]
```diff
@@ -17,24 +15,19 @@ import {
 
 interface VotingPowerChartProps {
   validatorAddress: ValidatorAddr;
+  singleStakingDenom?: string;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/bonded-token-changes/VotingPowerChart.tsx:18]: Let's use Option to make it a required field while can be undefined

```
singleStakingDenom: Option<string>
```

File: src/lib/pages/validator-details/components/bonded-token-changes/VotingPowerChart.tsx [hunk 4]
```diff
@@ -17,24 +15,19 @@ import {
...
+  assetInfos?: AssetInfos;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/bonded-token-changes/VotingPowerChart.tsx:19]: ditto

File: src/lib/pages/validator-details/components/bonded-token-changes/index.tsx [hunk 5]
```diff
@@ -1,20 +1,28 @@
...
 import { RelatedTransactionTable } from "../tables/RelatedTransactionsTable";
-import type { ValidatorAddr } from "lib/types";
+import type { AssetInfos, ValidatorAddr } from "lib/types";
 
 import { VotingPowerChart } from "./VotingPowerChart";
 
 interface BondedTokenChangesProps {
   validatorAddress: ValidatorAddr;
+  singleStakingDenom?: string;
+  assetInfos?: AssetInfos;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/bonded-token-changes/index.tsx:11]: ditto

File: src/lib/pages/validator-details/components/bonded-token-changes/index.tsx [hunk 6]
```diff
@@ -1,20 +1,28 @@
...
 }
 
 export const BondedTokenChanges = ({
   validatorAddress,
+  singleStakingDenom,
+  assetInfos,
 }: BondedTokenChangesProps) => {
   return (
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/bonded-token-changes/index.tsx:19]: Make it implicit return

File: src/lib/pages/validator-details/components/validator-overview/ValidatorDescription.tsx [hunk 7]
```diff
@@ -1,15 +1,25 @@
...
+    <Text variant="body2" fontWeight={500} as="h6" color="text.dark">
       Validator Description
-    </Heading>
-    <Text variant="body1" color="text.main">
-      Lorem ipsum, dolor sit amet consectetur adipisicing elit. Ullam ad vel
-      assumenda, temporibus maiores quidem quibusdam sapiente expedita officiis
-      minus! Officiis itaque nisi reprehenderit tempora eaque eos quam, in
-      excepturi.
+    </Text>
+    <Text variant="body2" color="text.main">
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-overview/ValidatorDescription.tsx:21]: The template before you replaced use body1, please double check this

File: src/lib/pages/validator-details/components/validator-overview/index.tsx [hunk 8]
```diff
@@ -18,61 +18,68 @@ interface ValidatorOverviewProps {
   onSelectVotes: () => void;
   onSelectPerformance: () => void;
   onSelectBondedTokenChanges: () => void;
+  isActive: boolean;
+  isJailed: boolean;
+  details: string;
   validatorAddress: ValidatorAddr;
+  singleStakingDenom?: string;
+  assetInfos?: AssetInfos;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-overview/index.tsx:26]: Remove ? And use Option

File: src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx [hunk 9]
```diff
@@ -13,23 +26,67 @@ const StatWithLabel = ({ label, value }: { label: string; value: string }) => (
...
-    <StatWithLabel label="Estimated APR" value="6.21%" />{" "}
-    <DotSeparator bg="gray.600" display={{ base: "none", md: "flex" }} />
-    <StatWithLabel label="Delegators" value="253" />
-  </Flex>
-);
+interface ValidatorStatsProps {
+  validatorAddress: ValidatorAddr;
+  commissionRate: number;
+  totalVotingPower: Big;
+  singleStakingDenom?: string;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx:33]: Option

File: src/lib/pages/validator-details/components/validator-overview/index.tsx [hunk 10]
```diff
@@ -18,61 +18,68 @@ interface ValidatorOverviewProps {
...
+            />
+          )}
+        </>
+      )}
+      {isMobile && !move.enabled && (
+        <BondedTokenChangeMobileCard
+          denom="OSMO"
+          onViewMore={onSelectBondedTokenChanges}
+        />
+      )}
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-overview/index.tsx:69]: remove all `move.enabled` condition

```
{
    isMobile ? (
        <BondedTokenChangeMobileCard
          denom="OSMO"
          onViewMore={onSelectBondedTokenChanges}
        />
    ) : (
        <>
          <Flex backgroundColor="gray.900" p={6} rounded={8} w="100%">
            <RecentBlocksSection hasTitle />
          </Flex>
          <VotingPowerChart
            validatorAddress={validatorAddress}
            singleStakingDenom={singleStakingDenom}
            assetInfos={assetInfos}
          />
        </>
    )
}
```

File: src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx [hunk 11]
```diff
@@ -1,8 +1,21 @@
...
+import type { ValidatorAddr } from "lib/types";
+import { formatPrettyPercent } from "lib/utils";
 
-const StatWithLabel = ({ label, value }: { label: string; value: string }) => (
+const StatWithLabel = ({
+  label,
+  value,
+}: {
+  label: string;
+  value: string | ReactNode;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx:17]:
```
label: string;
value: Option<string>;
isLoading: boolean;
```

and modify line 23 to
```
{
    isLoading ? (
        <Spinner size="sm" />
    ) : (
        <Text variant="body2" fontWeight={600} color={value ? "text.main" : "text.dark"}>
            {value ?? "N/A"}
        </Text>
    )
}
```

File: src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx [hunk 12]
```diff
@@ -13,23 +26,67 @@ const StatWithLabel = ({ label, value }: { label: string; value: string }) => (
...
+  validatorAddress,
+  commissionRate,
+  totalVotingPower,
+  singleStakingDenom,
+}: ValidatorStatsProps) => {
+  const { data: stakingProvisions } = useValidatorStakingProvisions();
+  const { data: delegations, isLoading: isDelegationsLoading } =
+    useValidatorDelegators(validatorAddress);
+
+  if (!stakingProvisions) return <Spinner size="sm" />;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx:46]: remove

File: src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx [hunk 13]
```diff
@@ -13,23 +26,67 @@ const StatWithLabel = ({ label, value }: { label: string; value: string }) => (
...
+  singleStakingDenom?: string;
+}
+
+export const ValidatorStats = ({
+  validatorAddress,
+  commissionRate,
+  totalVotingPower,
+  singleStakingDenom,
+}: ValidatorStatsProps) => {
+  const { data: stakingProvisions } = useValidatorStakingProvisions();
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx:42]: `const { data: stakingProvisions, isLoading: isStakingProvisionsLoading }`

File: src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx [hunk 14]
```diff
@@ -13,23 +26,67 @@ const StatWithLabel = ({ label, value }: { label: string; value: string }) => (
...
+        totalVotingPower.mul(1 - commissionRate)
+      )
+    : undefined;
+
+  const delegationsValue =
+    !isDelegationsLoading && delegations ? (
+      String(delegations.total)
+    ) : (
+      <Spinner size="sm" />
+    );
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx:59]:
```
const delegatorsCount = delegations ? (
      String(delegations.total)
    ) : undefined;
```

File: src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx [hunk 15]
```diff
@@ -13,23 +26,67 @@ const StatWithLabel = ({ label, value }: { label: string; value: string }) => (
...
+  const { data: delegations, isLoading: isDelegationsLoading } =
+    useValidatorDelegators(validatorAddress);
+
+  if (!stakingProvisions) return <Spinner size="sm" />;
+
+  const estimatedApr = singleStakingDenom
+    ? stakingProvisions.stakingProvisions.div(
+        totalVotingPower.mul(1 - commissionRate)
+      )
+    : undefined;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx:52]:
```
const estimatedApr = stakingProvisions
    ? formatPrettyPercent(stakingProvisions.stakingProvisions.div(
        totalVotingPower.mul(1 - commissionRate)
      ), 2, true)
    : undefined;
```

File: src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx [hunk 16]
```diff
@@ -13,23 +26,67 @@ const StatWithLabel = ({ label, value }: { label: string; value: string }) => (
...
+      mt={{ base: 2, md: 1 }}
+      mb={{ base: 0, md: 1 }}
+      px={{ base: 3, md: 0 }}
+      py={{ base: 1, md: 0 }}
+    >
+      <StatWithLabel
+        label="Commission"
+        value={formatPrettyPercent(commissionRate, 2, true)}
+      />
+      {estimatedApr && (
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx:79]:
```
{
    singleStakingDenom && (
        ...
    )
}
```

File: src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx [hunk 17]
```diff
@@ -13,23 +26,67 @@ const StatWithLabel = ({ label, value }: { label: string; value: string }) => (
...
+      borderRadius={4}
+      mt={{ base: 2, md: 1 }}
+      mb={{ base: 0, md: 1 }}
+      px={{ base: 3, md: 0 }}
+      py={{ base: 1, md: 0 }}
+    >
+      <StatWithLabel
+        label="Commission"
+        value={formatPrettyPercent(commissionRate, 2, true)}
+      />
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx:78]:
```
value={formatPrettyPercent(commissionRate, 2, true)}
isLoading={false}
```

File: src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx [hunk 18]
```diff
@@ -13,23 +26,67 @@ const StatWithLabel = ({ label, value }: { label: string; value: string }) => (
...
+      {estimatedApr && (
+        <>
+          <DotSeparator bg="gray.600" display={{ base: "none", md: "flex" }} />
+          <StatWithLabel
+            label="Estimated APR"
+            value={formatPrettyPercent(estimatedApr.toNumber(), 2, true)}
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx:84]:
```
value={estimatedApr}
isLoading={isStakingProvisionsLoading}
```

File: src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx [hunk 19]
```diff
@@ -13,23 +26,67 @@ const StatWithLabel = ({ label, value }: { label: string; value: string }) => (
...
+          />
+        </>
+      )}
+      <DotSeparator bg="gray.600" display={{ base: "none", md: "flex" }} />
+      <StatWithLabel label="Delegators" value={delegationsValue} />
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorStats.tsx:89]:
```
value={delegatorsCount}  
isLoading={isDelegationsLoading}  
```

File: src/lib/pages/validator-details/components/validator-top/ValidatorTitle.tsx [hunk 20]
```diff
@@ -21,7 +29,7 @@ export const ValidatorTitle = () => (
           variant={{ base: "h6", md: "h5" }}
           color="accent.main"
         >
-          #111
+          #{info.rank}
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorTitle.tsx:32]: don't render this `Heading` if rank is `null`

File: src/lib/pages/validator-details/components/validator-top/index.tsx [hunk 21]
```diff
@@ -1,15 +1,27 @@
...
-import { ValidatorImage } from "./ValidatorImage";
 import { ValidatorStats } from "./ValidatorStats";
 import { ValidatorTitle } from "./ValidatorTitle";
 import { WebsiteButton } from "./WebsiteButton";
 
-export const ValidatorTop = () => (
+interface ValidatorTopProps {
+  info: ValidatorData;
+  totalVotingPower: Big;
+  singleStakingDenom?: string;
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/index.tsx:17]: `Option<string>`

File: src/lib/types/validator.ts (hunk 1 above)
Comment (songwongtp) [inline src/lib/types/validator.ts:26]: Remove

**Review by songwongtp — CHANGES_REQUESTED** [review]

File: src/lib/pages/validator-details/components/validator-overview/ValidatorDescription.tsx (hunk 7 above)
Comment (Poafs1) [reply src/lib/pages/validator-details/components/validator-overview/ValidatorDescription.tsx:21]: Yes, both are using font size `14px` which is `body2` with font weight `500` for header and `400` as normal for body.

File: src/lib/pages/validator-details/components/validator-overview/index.tsx [hunk 22]
```diff
@@ -18,64 +18,67 @@ interface ValidatorOverviewProps {
...
+  return isActive && !isJailed ? (
+    <Flex direction="column" gap={{ base: 4, md: 6 }} pt={6}>
+      <ValidatorDescription details={details} />
+      <Flex gap={{ base: 4, md: 6 }} direction={{ base: "column", md: "row" }}>
+        <VotingPowerOverview />
+        <UptimeSection onViewMore={onSelectPerformance} />
       </Flex>
+      {isMobile ? (
+        <BondedTokenChangeMobileCard
+          denom="OSMO"
```
[the remark this answers is not in the document]
Comment (Poafs1) [reply src/lib/pages/validator-details/components/validator-overview/index.tsx:51]: Noted, will fix accordingly with next PR about Bounded Token

**Review by evilpeach — APPROVED** [review]

File: src/lib/pages/validator-details/components/validator-top/ValidatorTitle.tsx [hunk 23]
```diff
@@ -1,38 +1,48 @@
...
-        <Heading
-          as="h5"
-          mt={{ base: 1, md: 0 }}
-          ml={{ base: 1, md: 0 }}
-          variant={{ base: "h6", md: "h5" }}
-          color="accent.main"
-        >
-          #111
-        </Heading>
+        {info?.rank && (
```
Comment (songwongtp) [inline src/lib/pages/validator-details/components/validator-top/ValidatorTitle.tsx:25]: remove ?

**Review by songwongtp — APPROVED** [review]

**Pull request merged**